accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -2746,45 +2746,75 @@ To match a pattern `p` against a value `v`:
         *This type test may get elided. See "Pointless type tests and legacy
         types" below.*
 
-    2.  Let `l` be the length of the list determined by calling `length` on `v`.
-
-    3.  Let `h` be the number of non-rest elements preceding the rest element if
+    2.  Let `h` be the number of non-rest elements preceding the rest element if
         there is one, or the number of elements if there is no rest element.
 
-    4.  Let `t` be the number of non-rest elements following the rest element if
-        there is one, or zero otherwise.
+    3.  Let `t` be the number of non-rest elements following the rest element if
+        there is one, or `0` otherwise.
+
+    4.  Check the length. If `p` is empty or has any non-rest elements:
+
+        1.  Let `l` be the length of the list determined by calling `length` on
+            `v`. *We only call `length` on the list if needed.*
+
+        2.  If `p` has both a rest element and `h + t > 0`:
+
+            1.  If `l < h + t` then the match fails.
+
+            *When there are non-rest elements and a rest element, the list must
+            be at least long enough to match the non-rest elements.*
+
+        3.  Else if `h + t > 0` *(and `p` has no rest element)*:
 
-    3.  If `p` has no rest element and `l` is not equal to `h` then the match
-        fails. If `p` has a rest element and `l` is less than `h + t` then the
-        match fails. *These match failures become runtime exceptions if the list
-        pattern is in an irrefutable context.*
+            1.  If `l != h + t` then the match fails.
 
-    4.  Match the head elements. For `i` from `0` to `h - 1`, inclusive:
+            *If there are only non-rest elements, then the list must have
+            exactly the same number of elements.*
 
-        1.  Extract the element value `e` by calling `[]` on `v` with index `i`.
+        4.  Else `p` is empty:
```

Comment:
Whoops, I tried to comment on this line earlier but I guess I forgot to click the button and my draft got deleted.

We're not covering the case where `p` is `[...]`.  I think we could fix this by saying:

4.  Else *(`p` is empty or contains just a single "rest" pattern)*:
    1. If `l > 0` and `p` is empty, then the match fails.

Reply:
Better still, swap the order and make it "If `p` is empty and `l > 0`", that way it's clear that if `p` is `[...]`, we don't do a length check.